### PR TITLE
 feat: await_block LRU cache (non-global) 

### PR DESCRIPTION
**Pull request by dpc**
On top of #4035 and #4042.

This is a version #4046 without global statics.

Wasm couldn't handle `lazy_static`, and globals are icky. The price is wrangling with an already overly complicated set of traits in this area.

### CODE REVIEW

File: fedimint-core/src/api.rs [hunk 1]
```diff
@@ -360,15 +385,6 @@ pub trait GlobalFederationApi {
 
     async fn await_transaction(&self, txid: TransactionId) -> FederationResult<TransactionId>;
 
-    async fn await_output_outcome<R>(
```
Comment (dpc) [inline fedimint-core/src/api.rs:363]: Not object safe so had be kicked out (to `impl DynGlobalApi`).

File: fedimint-core/src/api.rs [hunk 2]
```diff
@@ -440,6 +474,78 @@ where
...
+        let mut block_lock = block_entry_arc.lock().await;
+
+        if let Some(cached_res) = block_lock.clone() {
+            debug!(block_index, "Using a cached await_block response");
+            return Ok(cached_res);
+        }
+
+        let block = self.await_block_raw(block_index, decoders).await?;
+
+        *block_lock = Some(block.clone());
```
Comment (elsirion) [inline fedimint-core/src/api.rs:545]: Should we also debug log wen we are fetching the block or is that already done by `await_block_raw`?

**Review by elsirion — COMMENTED** [review]
Looks good, but blocked on two other PRs

File: fedimint-core/src/api.rs (hunk 2 above)
Comment (dpc) [reply fedimint-core/src/api.rs:545]: Will check when I rebase, it should also be easier to review in full.

Comment (dpc) [reply fedimint-core/src/api.rs:603]: Unclear if it's really needed.

**Comment by dpc** [comment]
@elsirion :heavy_check_mark: :question: :pray: :rofl:

File: fedimint-core/src/api.rs (hunk 2 above)
Comment (elsirion) [reply fedimint-core/src/api.rs:603]: I mean, you do have the log when the block/session was cached. But fine by me :man_shrugging: more logging more good though

**Review by elsirion — APPROVED** [review]

File: fedimint-core/src/api.rs [hunk 3]
```diff
@@ -497,6 +544,78 @@ where
...
+        let block_entry_arc = lru_lock
+            .get_or_insert(block_index, || Arc::new(tokio::sync::Mutex::new(None)))
+            .clone();
+
+        // we drop the lru lock so requests for other `block_index` can work in parallel
+        drop(lru_lock);
+
+        // but multiple request for the same `block_index` will serialize on the
+        // per-index lock, avoiding doing the same work
+        let mut block_lock = block_entry_arc.lock().await;
```
Comment (Maan2003) [inline fedimint-core/src/api.rs:594]: block_entry can be a `OnceCell`

https://docs.rs/tokio/latest/tokio/sync/struct.OnceCell.html

**Review by Maan2003 — APPROVED** [review]

File: fedimint-core/src/api.rs (hunk 2 above)
Comment (Maan2003) [reply fedimint-core/src/api.rs]: there is already log in await_block_raw

**Review by Maan2003 — APPROVED** [review]

File: fedimint-core/src/api.rs (hunk 3 above)
Comment (Maan2003) [reply fedimint-core/src/api.rs]: i pushed https://github.com/fedimint/fedimint/pull/4080/commits/4efb0f22a261e007ed2d42d71a2fca60b9a32d16 to reduce review cycles

**Review by Maan2003 — APPROVED** [review]

**Review by elsirion — APPROVED** [review]
Good catch, makes the code a lot simpler!

File: fedimint-core/src/api.rs (hunk 3 above)
Comment (dpc) [reply fedimint-core/src/api.rs]: Interesting.

**Pull request merged**